### PR TITLE
feat: update the response code to build short response snippets

### PR DESCRIPTION
**Pull request by ozym**
This helps build configurations that simply needs sensitivities and gains etc.

Most of this code has been moved from internal/stationxml.

The main type InstrumentResponse is likely to be renamed once the current Response struct code is removed.

### CODE REVIEW

**Review by jajera — DISMISSED** [review]
Looks good on a platform perspective

File: resp/instrument.go [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+}
+
+// Preamp is used to adjusts the datalogger gains, this is in addition to the default values.
+func Preamp(preamp float64) InstrumentResponseOpt {
+	return func(r *InstrumentResponse) {
+		r.Preamp = preamp
+	}
+}
+
+// NewResponse builds a Response with the given options.
```
Comment (wilsonjord) [inline resp/instrument.go:89]: nickpick: `NewInstrumentResponse builds`

File: resp/resp.go [hunk 2]
```diff
@@ -24,7 +26,7 @@ func LookupFS(fsys fs.FS, response string) ([]byte, error) {
 
 		// return the first one found
 		for _, name := range names {
-			data, err := fs.ReadFile(files, name)
+			data, err := fs.ReadFile(fsys, name)
```
Comment (wilsonjord) [inline resp/resp.go:29]: question: don't quite understand this, is this a bugfix?

File: resp/response_type.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+	Zeros []PoleZero `xml:"Zero"`
+	Poles []PoleZero `xml:"Pole"`
+}
+
+// Gain ccalculates the poles and zeros response gain at a given frequency
+func (pz PolesZerosType) Gain(freq float64) float64 {
+
+	var w complex128
+	switch pz.PzTransferFunctionType {
```
Comment (wilsonjord) [inline resp/response_type.go:106]: nitpick(non-blocking): I believe `switch` is recommend for 3 or more conditions, and `if` for simply binary.

File: resp/resp.go (hunk 2 above)
Comment (ozym) [reply resp/resp.go:29]: Ah, yes, realised after testing that the variable was being shadowed.

File: resp/response_type.go (hunk 3 above)
Comment (ozym) [reply resp/response_type.go:106]: This is my c programming background protesting about the lack of a conditional statement in go.  I've tended to have an aversion to if else blocks and have standardised on a switch.

Comment (wilsonjord) [reply resp/response_type.go:106]: I understand :-) I didn't realise how much I missed conditional operators until I started using Go...

**Review by wilsonjord — APPROVED** [review]
lgtm

**Comment by salichon** [comment]
From the "DMC" side then  finalizing review 
- Local  branch download have a look ..  
- Compile and run the new code
- Compare/validate the output of the new branch  (use stationxml-build code )   

Does it sound like (about) a plan @ozym

**Comment by ozym** [comment]
> From the "DMC" side then finalizing review
> 
> * Local  branch download have a look ..
> * Compile and run the new code
> * Compare/validate the output of the new branch  (use stationxml-build code )
> 
> Does it sound like (about) a plan @ozym ?

This shouldn't impact anything (it's a library) but the closest may be to run the "test" in the /resp folder, e.g. (cd resp; go test).

**Comment by elidana** [comment]
FWIW, without knowing what I was doing and with @salichon we did the following:


- download a snippet from the NSF SAGE nominal response library
- saved into resp/files
- go test


first test fail, had to:
- update the schema reference (xml version at the top)
- reformat the xml
- the go test fails

I think the test fails only for formatting issues with the snippets, but with no change on the repo the test pass nicely.


below the details


```
$> curl -O 'http://service.iris.edu/irisws/nrl/1/combine?instconfig=sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel&format=stationxml-resp'
$> mv combine sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel.xml
```
edit xml to change header (xml version)

```
$> go test
--- FAIL: TestFiles (0.05s)
    --- FAIL: TestFiles/check_resp_file:_sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel.xml (0.02s)
        resp_test.go:80:   []uint8(
              	"""
              	<?xml version="1.0" encoding="UTF-8"?>
              	<Response>
              	  <InstrumentSensitivity>
            - 	    <Value>-1.50000E+03</Value>
            + 	    <Value>-1500</Value>
              	    <Frequency>1</Frequency>
              	    <InputUnits>
              	... // 6 identical lines
              	    </OutputUnits>
              	  </InstrumentSensitivity>
            - 	  <Stage number="1" resourceId="NRL:AnaPAZ_IUmPs_NZ8_NP13_PH8294398_AN1.02995ep13_FN1::NRL:Gain_SGm1500_FN1">
            - 	    <PolesZeros>
            + 	  <Stage number="1">
            + 	    <PolesZeros resourceId="">
              	      <InputUnits>
              	        <Name>m/s</Name>
              	... // 101 identical lines
              	"""
              )
            
FAIL
exit status 1
FAIL	github.com/GeoNet/delta/resp	1.633s
```


not sure how valid all of this is for a review

**Comment by ozym** [comment]
An interesting test, you're testing the different XML building software mechanisms. e.g. 1500 vs 1.5E03 but the resourceId one is more interesting.  This should either copy as is, or ignore.

**Comment by ozym** [comment]
@elidana @salichon this small update should change the output of test for this file. It won't fix the 1.5e+03 vs 1500 issue as that's related to the coding language.

**Comment by ozym** [comment]
We could also look at adding the xsd  element, e.g.

```
 xmllint --schema 'http://ds.iris.edu/files/xml/station/fdsn-station-response-1.1.xsd' --noout  files/sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel.xml
files/sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel.xml validates
```

but that could be an extra step for manual validation.

**Comment by elidana** [comment]
Thanks Mark!
I did updated local branch and repeat the steps. The NRL response file needs a bit of love before the test can pass (steps below details what) - @salichon we might want to keep those somewhere in our wiki if that needs to be done again?

1. download from NRL (with curl command below, file is named `combine`)
1. a bunch of manual edits on the NRL xml library (reformat and manual changes)
1. and the test pass!


```
curl -O 'http://service.iris.edu/irisws/nrl/1/combine?instconfig=sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel&format=stationxml-resp'
xmllint --format combine > files/sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel.xml 
dos2unix files/sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel.xml
vi files/sensor_Streckeisen_STS-5A_LP120_SG1500_STgroundVel.xml 
```
.... (see below for manual edits)
```
$> go test
PASS
ok  	github.com/GeoNet/delta/resp	1.602s
```
history



```diff
@@ -1,7 +1,7 @@
-<?xml version="1.0"?>
-<Response xmlns="http://www.fdsn.org/xml/station/1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.fdsn.org/xml/station/1 http://ds.iris.edu/files/xml/station/fdsn-station-response-1.1.xsd">
+<?xml version="1.0" encoding="UTF-8"?>
+<Response>
   <InstrumentSensitivity>
-    <Value>-1.50000E+03</Value>
+    <Value>-1500</Value>
     <Frequency>1</Frequency>
     <InputUnits>
       <Name>m/s</Name>
```

**Review by elidana — APPROVED** [review]
from the little I understand , approved!!

**Pull request merged**